src/hotspot/share/gc/shared/gcLocker.cpp

```diff
@@ -123,33 +106,30 @@ bool GCLocker::should_discard(GCCause::Cause cause, uint total_collections) {
          (_total_collections != total_collections);
 }
 
-void GCLocker::jni_lock(JavaThread* thread) {
+void GCLocker::stall_until_no_needs_gc(JavaThread* thread) {
   assert(!thread->in_critical(), "shouldn't currently be in a critical region");
   MonitorLocker ml(JNICritical_lock);
   // Block entering threads if there's a pending GC request.
   while (needs_gc()) {
     // There's at least one thread that has not left the critical region (CR)
     // completely. When that last thread (no new threads can enter CR due to the
-    // blocking) exits CR, it calls `jni_unlock`, which sets `_needs_gc`
+    // blocking) exits CR, it calls `unlock_critical`, which sets `_needs_gc`
     // to false and wakes up all blocked threads.
-    // We would like to assert #threads in CR to be > 0, `_jni_lock_count > 0`
-    // in the code, but it's too strong; it's possible that the last thread
-    // has called `jni_unlock`, but not yet finished the call, e.g. initiating
+    // We would like to assert #threads in CR to be > 0, i.e. `_jni_lock_count > 0`,
+    // but that would be too strong -- it's possible that the last thread
+    // has called `unlock_critical`, but not yet finished the call, e.g. initiating
     // a GCCause::_gc_locker GC.
     ml.wait();
   }
-  thread->enter_critical();
-  _jni_lock_count++;
-  increment_debug_jni_lock_count();
 }
 
-void GCLocker::jni_unlock(JavaThread* thread) {
-  assert(thread->in_last_critical(), "should be exiting critical region");
-  MutexLocker mu(JNICritical_lock);
-  _jni_lock_count--;
-  decrement_debug_jni_lock_count();
-  thread->exit_critical();
-  if (needs_gc() && !is_active_internal()) {
+void GCLocker::ack_needs_gc_if_last_thread(JavaThread* thread) {
+  assert(!thread->in_critical(), "shouldn't currently be in a critical region");
+  assert(needs_gc(), "precondition");
+  jint old_value = Atomic::fetch_and_add(&_jni_lock_count, -1);
```

Comment:
Even assuming it's okay to modify _jni_lock_count outside the lock (which I'm not
at all sure about), I'm not convinced it's an improvement to do so.